src/main.rs

```diff
@@ -53,13 +53,25 @@ fn run() -> Result<(), MainError> {
             }
             let manifest_path = cwd.join("package.json");
             let manifest = read_package_manifest(&manifest_path)?;
-            if let Some(command) = manifest.scripts.get(&args.script) {
-                eprintln!("> {command}");
-                run_script(&args.script, command, &cwd)
+            if let Some(name) = args.script {
+                if let Some(command) = manifest.scripts.get(&name) {
+                    eprintln!("> {command}");
+                    run_script(&name, command, &cwd)
+                } else {
+                    PnError::MissingScript { name }
+                        .pipe(MainError::Pn)
+                        .pipe(Err)
+                }
+            } else if manifest.scripts.is_empty() {
+                println!("There are no scripts in package.json");
+                Ok(())
             } else {
-                PnError::MissingScript { name: args.script }
-                    .pipe(MainError::Pn)
-                    .pipe(Err)
+                println!("Commands available via `pn run`:");
+                for (name, command) in manifest.scripts {
+                    println!("  {name}");
+                    println!("    {command}");
+                }
```

Comment:
It doesn't make a big difference in this case but as far as I know it is faster to first combine the output to one string and the print it with one `println`. Let's change it to use best practices.

Reply:
If it doesn't make a big difference, then readability should be preferable in my opinion.

The reason I chose to write `println!` twice because I'd like two lines being on two lines.

But if you still disagree with me, then you may apply this suggestion manually.

```suggestion
                println!("Commands available via `pn run`:");
                for (name, command) in manifest.scripts {
                    println!("  {name}\n    {command}");
                }
```

And if you change your opinion, please just mark this as resolved.

Reply:
No, I mean all lines concatenated to a string and then printed out
like here

https://github.com/pnpm/pnpm/blob/c686768f060520b42b51dc1d1e382e4a25eb5ac3/exec/plugin-commands-script-runners/src/run.ts#L296C1-L314

Reply:
> https://github.com/pnpm/pnpm/blob/c686768f060520b42b51dc1d1e382e4a25eb5ac3/exec/plugin-commands-script-runners/src/run.ts#L296C1-L314

That is overly verbose in my opinion, and for a little gain.

You got away with it in JavaScript because V8's string is actually a [rope](https://en.wikipedia.org/wiki/Rope_(data_structure)) which is optimized for string concatenation.

Rust's string, however, is just an array of characters. Copying strings then print is *worse* than just print static strings multiple times.

Reply:
That being said, there are still way to achieve the "best of both worlds": https://docs.rs/text-block-macros/0.1.1/text_block_macros/

It will guarantee best possible performance (better than even V8's rope string), whilst still keeps 2 lines in exactly 2 lines.

Let me know if you would want to use it?

Reply:
It seems that I misunderstood what you meant.

Anyway, I created https://github.com/pnpm/pn/issues/41, proper benchmark is probably needed.